examples/train-gpt.py

```diff
@@ -11,14 +12,66 @@
 d_value = 32
 num_blocks = 4
 
+# Llama-7b-like values, excluding the vocabulary size.
+vocab_size = 256
+context = 1024
+num_heads = 32
+d_embed = 4096
+d_query = 128
+d_value = 128
+num_blocks = 4
+
+GPU_16BIT_FLOPS = {
+    "h100-sxm": 1.979e15 / 2,
+    "h100-pcie": 1.513e15 / 2,
+    "a100": 312e12,
+    "v100-sxm": 125e12,
+    "6000A": 364.25e12,
+    "4090": 165.2 * 10**12,
+    "3090": 71 * 10**12,
```

Comment:
Is this correct? Wikipedia only lists the 2:1 sparse TFLOPs but the [source included in the article](https://web.archive.org/web/20220112113553/https://www.anandtech.com/show/16057/nvidia-announces-the-geforce-rtx-30-series-ampere-for-gaming-starting-with-rtx-3080-rtx-3090) shows it as 143 TFLOPs. I think that would explain why you're seeing high MFU. On an A10 (with `num_blocks = 3` because of memory spike when adam initialises, `batch_size=8`) I see:
```
...
step: 100        train loss: 2.77        test loss: 2.81         tokens/gpu/sec: 14982.88  MFU: 43.46%
```
Assuming an A10 has peak 16 bit 125 TFLOPs. This is pretty closely in line with what I've seen with nanoGPT experiments. The code matches nanoGPT closely so that would make sense.

Reply:
Please always consult the appropriate [nvidia whitepaper](https://images.nvidia.com/aem-dam/en-zz/Solutions/geforce/ampere/pdf/NVIDIA-ampere-GA102-GPU-Architecture-Whitepaper-V1.pdf) when looking for GPU performance numbers. if you have issues finding these PDFs in the future, I have [developed a simple frontend](https://152334h.github.io/gpu-table/) that aggregates a number of manually collected sources.

Regarding the observed MFU of 3090,
1. all gamer gpus have crippled fp32 accumulation such that the effective performance of any tensor core is half of what it should be. see the table from the link above:
     ![image](https://github.com/jxbz/modula/assets/54623771/fdf59e09-bece-4609-9f66-bede273a7739)
2. all pytorch matmuls use fp32 accumulation for all GEMMs; this is [impossible to change at runtime](https://x.com/main_horse/status/1742013129650065861), so all gamer GPUs on pytorch get half the performance of what they would with fp16 accumulation.
3. because of this disadvantage, gamer GPUs also have a much lower FLOPs-to-membandwidth ratio than datacenter GPUs do, and it is much easier to achieve high MFU on them than on their corresponding datacenter equivalents.

I hope this addresses your concerns.

Reply:
OK so if I use a triton matmul to accumulate in fp16 I should see the MFU on the A10 increase to around what you're seeing with a 3090?

Reply:
The A10 has equivalent fp16 vs fp32 accum performance, unlike the 3090. This means that the A10 will always appear to have lower MFU relative to the 3090, regardless of accumulation precision.

Please note that the assumed peak tflops used in MFU calculation is based on the reported spec sheet values for fp32 accumulation. You may interpret this as the 3090 having an "easier target" to accomplish.

Reply:
How about this for a "lazy coder's MFU": measure the FLOP ceiling using just PyTorch matmuls and assume that's the peak I can get without doing extra work in triton or C++. I wrote [this script](https://gist.github.com/gaviag-cerebras/5dd1fa407077a3728acc622d33438621) and it measures the A10 at 69 TFLOPs, then the results are pretty much exactly what you see on the 3090. Out of interest, what flop ceiling do you see on the 3090 with the same script?

Reply:
I've done the simple lazy coder's approach in the past; it's usually within 95% of the spec on gamer GPUs. These are my 3090 results, which are not unexpected:
```
Running on device: cuda
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa483f67b20>
FLOP Ceiling Measurement: Precision: bfloat16
Matrix size: 32768x32768
setup: from __main__ import matrix1, matrix2, matmul
  Median: 1.08 s
  IQR:    0.00 s (1.08 to 1.08)
  10 measurements, 1 runs per measurement, 64 threads
FLOP Ceiling: 65.30 TFLOP/s
```

However, **your observation of the A10 achieving only 69TFLOPs surprised me**, so I spun up A10 (Lambdalabs) && A10g (AWS) instances to [test triton fp16 vs torch fp32 accum](https://gist.github.com/152334H/25e81ff7811b7f7fb049e23e756cfe80).

# A10
![image](https://github.com/jxbz/modula/assets/54623771/4fbeff11-2d5b-474c-909f-296cc6c6b193)

I am also able to replicate your script's results:
```
Running on device: cuda
<torch.utils.benchmark.utils.common.Measurement object at 0x7f3bfe2635b0>
FLOP Ceiling Measurement: Precision: bfloat16
Matrix size: 32768x32768
setup: from __main__ import matrix1, matrix2, matmul
  Median: 1.01 s
  IQR:    0.07 s (0.98 to 1.05)
  99 measurements, 1 runs per measurement, 30 threads
FLOP Ceiling: 69.80 TFLOP/s
```
During the execution of your script, I noticed that **the A10's GPU clock speed in nvtop is drastically limited at 100% utilization**. When initially boosted, it goes up close to 1700MHz, but after a period of continuous execution, it settles around 1000MHz.

#### **I believe that the A10's 150W TDP explains the reduced performance of the A10 relative to the expected 125TF spec**.

# A10g
There is a separate [spec sheet](https://d1.awsstatic.com/product-marketing/ec2/NVIDIA_AWS_A10G_DataSheet_FINAL_02_17_2022.pdf) for the A10g, which indicates it having **only 70 peak TFlops** and a TDP of 300W.

![image](https://github.com/jxbz/modula/assets/54623771/5eb1064f-e167-4c33-96bf-f2413ce1d1c9)

On 300W, it achieves this much:
![image](https://github.com/jxbz/modula/assets/54623771/23e7dac8-8f95-4c7f-8612-cf604056c8d1)

Limited to 150W, it performs as so:
![image](https://github.com/jxbz/modula/assets/54623771/eae566d7-fba7-4929-896a-3ab0b6dfe9c1)